url0e7d57473a_kyorohiro_HetimaUtil_tgz-pJ8-net_hetimatan_net_http_HttpServer3xxJ8/src/net/hetimatan/util/http/HttpGetRequestUri.java

```diff
@@ -0,0 +1,344 @@
+
+import net.hetimatan.io.file.MarkableFileReader;
+import net.hetimatan.io.file.MarkableReader;
+import net.hetimatan.io.file.MarkableReaderHelper;
+import net.hetimatan.io.filen.CashKyoroFile;
+
+// GET request
+// http://www.w3schools.com/tags/ref_httpmethods.asp
+// Request-URI    = "*" | absoluteURI | abs_path
+public class HttpGetRequestUri extends HttpObject {
+	private String mPath = "";
+	private String mHost = "";
+	private int mPort = 80;
+
+	private LinkedHashMap<String, String> mValues = new LinkedHashMap<String, String>();
+
+	// todo throw IOEXception?
+	public static HttpGetRequestUri crateHttpGetRequestUri(String requestPath) {
+		try {
+			CashKyoroFile base = new CashKyoroFile(requestPath.getBytes());
+			MarkableReader reader  = new MarkableFileReader(base, 100);
+			return HttpGetRequestUri.decode(reader);
+		} catch(Exception e) {
+			return new HttpGetRequestUri(requestPath);
+		}
+	}
+
+	public HttpGetRequestUri(String path) {
+		mPath = path;
+	}
+
+	@Override
+	public String toString() {
+		try {
+			String ret = HttpObject.createEncode(this);
+			return ret;
+		} catch (IOException e) {
+			return "#Failed toString#";
+		}
+	}
+
+	public void setHost(String host) {
+		mHost = host;
+	}
+
+	public void setPort(int port) {
+		mPort = port;
+	}
+
+	public void putVale(String key, String value) {
+		mValues.put(key, value);
+	}
+
+	public Set<String> keySet() {
+		return mValues.keySet();
+	}
+
+	//
+	// todo
+	public void setPath(String path) {
+		if(path.startsWith("/")) {
+			mPath = path;
+		} else {
+			mPath = "/"+path;
+		}
+	}
+
+	public String getPath() {
+		return mPath;
+	}
+
+	public String getValue(String key) {
+		return mValues.get(key);
+	}
+
+	private boolean mAbsPath = true;
+	public void IsAbsolutePath(boolean on) {
+		mAbsPath = !on;
+	}
+	//
+	// /test/demo_form.asp?name1=value1&name2=value2
+	//
+	@Override
+	public void encode(OutputStream output) throws IOException {
+		if(mAbsPath) {
+			encodeAbsPath(output);
+		} else {
+			encodeAbsolutePath(output);
+		}
+	}
+
+	public void encodeAbsolutePath(OutputStream output) throws IOException {
+		output.write(("http://"+mHost+":"+mPort).getBytes());
+		encodeAbsPath(output);
+	}
+
+	public void encodeAbsPath(OutputStream output) throws IOException {
+		output.write(mPath.getBytes());
+		Set<String> keys = mValues.keySet();
+		boolean isFirst = true;
+		for (String key : keys) {
+			if (true == isFirst) {
+				isFirst = false;
+				output.write("?".getBytes());
+			} else {
+				output.write("&".getBytes());
+			}
+			output.write(key.getBytes());
+			output.write("=".getBytes());
+			output.write(mValues.get(key).getBytes());
+		}
+	}
+
+	public static HttpGetRequestUri decode(String location) throws IOException {
+		MarkableFileReader reader = null;
+		try {
+			reader = new MarkableFileReader(location.getBytes());
+			HttpGetRequestUri geturi = HttpGetRequestUri.decode(reader);
+			return geturi;
+		} finally {
+			reader.close();
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 123:129, 336:342.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.